byway/src/main/webapp/recommendations.js

```diff
@@ -0,0 +1,254 @@
+  const myRoute = directionResult.routes[0];
+  for(leg of myRoute.legs) {
+    for(step of leg.steps) {
+      const avgLat = (step.start_location.lat() + step.end_location.lat()) / 2;
+      const avgLng = (step.start_location.lng() + step.end_location.lng()) / 2;
+      const avgLoc = {lat: avgLat, lng: avgLng};
+      if(step.distance.value > MIN_DISTANCE) {
+        regions.push(avgLoc);
+      }
+    }
+  }
+}
+
+/**
+ * Set a timeout to delay the browser.
+ * @param {Number} delayMs number of milliseconds
+ * @return empty promise after delayMs milliseconds passed
+ */
+function delayPromise(delayMs) {
+  return new Promise(resolve => setTimeout(resolve, delayMs));
+}
+
+/**
+ * Go through a user's interests and search for places
+ * fitting those interests. Search around the regions
+ * previously found using textSearch. Prioritize finding
+ * past results loaded through sessionStorage to avoid calling
+ * textSearch repeatedly from the PlacesService.
+ */
+async function loadRecommendations() {
+  for(interest of interests) {
+    for(region of regions) {
+      const request = {
+        location: region,
+        radius: RADIUS,
+        query: interest
+      }
+      const recommendationsSaved = sessionStorage.getItem(JSON.stringify(request));
+      if(recommendationsSaved !== null) {
+        addRecommendations(request, JSON.parse(recommendationsSaved));
+      } else{
+        // Prevent hitting the query limit from the maps API
+        // by setting a timeout in delayPromise
+        await delayPromise(250);
+        const placesFound = await findPlacesWithTextSearch(request);
+        if(placesFound !== undefined) {
+          addRecommendations(request, placesFound);
+        }
+      }
+    }
+  }
+}
+
+/**
+ * Use the textSearch function from PlacesService to find the results
+ * fitting the request object passed in. Convert callback function into
+ * a chain of promises to return the result directly. If the service
+ * returns a status other than OK, reject the promise and alert the user.
+ * @param {TextSearchRequest} request object with location, radius and query fields.
+ * @return promise either with PlaceResult[] results or undefined if rejected.
+ */
+function findPlacesWithTextSearch(request) {
+  return new Promise((resolve, reject) => {
+    placesService.textSearch(request, (result, status) => {
+      if(status === "OK") {
+        resolve(result);
+      } else {
+        reject(new Error(status));
+      }
+    });
+  }).catch((error) => {
+    alertUser(error.message);
+  });
+}
+
+/**
+ * Provides a message to the user if there was an issue
+ * with their request.
+ * TODO: Make a class to indicate as a "warning" and make text red.
+ * Also, organize on front end.
+ * @param {String} msgFromService status String from placesService
+ */
+function alertUser(msgFromService) {
+  let messageContainer = document.getElementById("message-container");
+  let message = document.createElement('p');
```

Comment:
Updated. I might try to find a better way to load the message on screen per unique status message, without having too add multiple "OVER_QUERY_LIMIT" messages or any other status. My update might still have this problem, since it is listing the statuses and there's no check if one already exists.